### ISSUE TITLE
Question: How to hardcode OrderBy in Function？

### ISSUE THREAD

**Issue by vincedan** [OData/WebApi#1957]
I posted question on [stackOverflow ](https://stackoverflow.com/questions/58771676/odata-webapi-how-to-hardcode-orderby-in-function?r=SearchResults), and nobody respond me by now. sorry for me to copy the post here, wish I could get help from community. 

I am using OData WebAPI (Microsoft.AspNet.OData). for some reasons, I need hardcore OrderBy in a function. for example and return like the following, 
` return DbContext.Companies.OrderByDescending(a => a.CompanyName);`
When I executed the function, orderby hardcoded did not work, I also checked sqlprofile, it seem framework did something and remove it. If I use $orderby queryoption, it is no problem.  btw,  it seems that filter conditions in Where clause in function will be combined with $filter option , that is well,  I need the same for orderby. Is there anyone knows how to solve it? 

thks in advance.

**Comment by vincedan** [comment]
Thank @NetTecture  for your response. I suggest OData team to implement it as new feature so that user could predefine orderby in function.

**Comment by NetTecture** [comment]
@vincedan It would have to happen at some sort of edm / model level ;) But yeah, I have been missing this quite long. Thanks a lot.

**Comment by KanishManuja-MS** [comment]
One workaround to implement is by making CompanyName a part of composite key and specifying Column attribute. I doubt that it will work without making the property part of the key and just using Column attribute but you can quickly try that out. 

One other workaround is to insert a orderby clause in ODataQueryOptions in your controller.

**Comment by NetTecture** [comment]
Given that the key is also the key for fast entity access this simply does not work. It also is non practical if the database is not particularly and only written for the odata access side.

**Comment by KanishManuja-MS** [comment]
@NetTecture Yes, you are correct. If for a schema it makes sense to have a composite key, first workaround can be chosen. Otherwise, the second approach would let you inject a custom order by in your controller action.

**Comment by vincedan** [comment]
Thank @KanishManuja-MS's response and your two workarounds. as @NetTecture said, my system is using Guid as PK for each entity and I need to add meaningful business logic in orderby.  In addition, the orderby rules are different based on entity composition(for example, a function may return entity A with orderby principal navigation which is entity B). LINQ expression is best solution for handling it.  And final user need to specify query options such as $orderby outside. all things is like lightswitch did.  Workaround 2 let me know a little essential in WebAPI,  modify orderByOptions of ODataQueryOptions and insert new OrderByNodes should achieve it but I guess it is not suitable for mass manipulation and provide framework/library solution should be better.

**Comment by TehWardy** [comment]
Is Options still a thing ?
Example from here ... 
https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/supporting-odata-query-options

```
public IQueryable<Product> Get(ODataQueryOptions opts)
{
    return opts.ApplyTo(products.AsQueryable())as IQueryable<Product>;
}
```

By passing the Options to be applied to the query in to every action and then creating an extension method you could modify the above to meet your needs with something like ...

```
public IQueryable<Product> Get(ODataQueryOptions opts)
{
    // add your order by logic to the options object
    opts.ApplySystemDefaults();

    // apply and return the result set 
    return opts.ApplyTo(products.AsQueryable())as IQueryable<Product>;
}
```

**Comment by vincedan** [comment]
thank @TehWardy response.
all I need is something like

`

        [HttpGet]
        [ODataRoute("GetCompanies3(EmployeeId={employeeId})")]
        public IQueryable<Company> GetCompanies3(int employeeId, ODataQueryOptions options)
        {
            return options.ApplyTo(db.Companies.Where(c => c.Employee.Integer2 == employeeId).OrderBy(c => c.Integer1)) as IQueryable<Company>;
        }

`
I am wondering orderby merging should be handled in odata library, the same as filters in Where clause.

**Comment by TehWardy** [comment]
did that work ?

**Comment by vincedan** [comment]
Of course not, or I will close this post.  :-)
if change options firstly, then invoke options.ApplyTo() does work.

**Comment by TehWardy** [comment]
have you tried constructing a new ODataQueryOptions instance and passing the stuff you DO want from the framework given one + your order by clause then doing ApplyTo

**Comment by vincedan** [comment]
@TehWardy  yes. I have tried it.  change ODataQueryOptions does work. as the following example, if client use $orderyby,  a special orderby column which predefined in the function will be inserted in existing options.
` 

       public IQueryable<Company> GetCompanies4(int employeeId, ODataQueryOptions options)
        {
            //return queryOptions.ApplyTo(db.Companies.Where(c => c.Employee.Integer2 == employeeId).OrderBy(c => c.Integer1)) as IQueryable<Company>;

            var source = db.Companies.Where(c => c.Employee.Integer2 == employeeId);

            if (options.OrderBy == null)
            {
                source = source.OrderBy(c => c.Integer1);
            }
            else
            {
                var entityType = options.Context.ElementType as IEdmEntityType;
                var ddProperty = entityType.DeclaredStructuralProperties().Single(p => p.Name == "Integer1");
                options.OrderBy.OrderByNodes.Insert(0, new OrderByPropertyNode(ddProperty, OrderByDirection.Ascending));
            }
            
            return options.ApplyTo(source) as IQueryable<Company>;
        }`
I have a many complex orderby predefined for lots of odata functions, as I mentioned before, I was wondering it should be handled in odataframework like filters in where clause.

**Comment by NetTecture** [comment]
he problem with this is that it only handles - and that seems what most people seem to imply is the normal case - the $orderby on the TOP LEVEL. Filters and ordering can be added to expands, too. Central handling or order in the model will allow to generate proper OrderBy also for expands. And yes, we do use that QUITE often.

**Comment by TehWardy** [comment]
What about ...

```
public IQueryable<Company> GetCompanies4(int employeeId, ODataQueryOptions options)
{
     var result = options.ApplyTo(db.Companies.Where(c => c.Employee.Integer2 == employeeId).OrderBy(c => c.Integer1)) as IQueryable<Company>;
     if(optons.OrderBy == null) { result = result.OrderBy(...); }
     return result;
}
```

**Comment by vincedan** [comment]
@TehWardy ,  invoke options.ApplyTo will replace existing orderby linq expression with [$orderby] in options. after that, if need to append orderby again to Queryable linq expression, it will replace all orderby logic again with new one. in addition, even it can be done as expected, I mean "append" orderby does work, the related business sequence for orderby is not correct. I guess [use expression trees to build dynamic queries](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/expression-trees/how-to-use-expression-trees-to-build-dynamic-queries).

**Comment by TehWardy** [comment]
Surely then the correct answer here is to update the calling query to include the expected OrderBy clause ... I mean ... that is literally what the param is for.
